cluster: Implement sanctions on pin leadership features on invalid license

IoannisRP opened the pull request:
Implements: [CORE-8042](https://redpandadata.atlassian.net/browse/CORE-8042) - Sanctions 18,19

Implement sanction(s) on leadership pinning properties when there is no valid license:
- If either the cluster config `default_leaders_preference` or a topic property `redpanda.leaders.preference` are set, the settings will be ignored.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

* none

[CORE-8042]: https://redpandadata.atlassian.net/browse/CORE-8042?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP commented:
- rebased on dev
- removed create topic / change topic configs which are split into separate PRs.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--1c4f0e07-617f-4a3c-b30c-105898ddc267-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/57859#01930da8-e2ef-4911-84cd-2359f0a22064
<!--/body-->

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--ffd275be-ea26-4901-92ea-831317748f86-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/57859#01930da7-368f-4289-8fac-b9216ccbbb47:
```
"rptest.tests.archive_retention_test.CloudArchiveRetentionTest.test_delete.cloud_storage_type=CloudStorageType.ABS.retention_type=retention.ms"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/57859-->
<!--3443ab77-5c55-4fa1-a208-843d442053fd-->
<!--title-->
## Retry command for [Build#57859](https://buildkite.com/redpanda/redpanda/builds/57859)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.ms"}
```
<!--/body-->

BenPope commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -821,8 +823,19 @@ leader_balancer_types::preference_index
+    if (is_sanctioned) {
+        vlog(
+          clusterlog.warn,
+          "A Redpanda Enterprise Edition license is required to use "
+          "enterprise feature \"default_leaders_preference\". "
+          "This property is being ignored.");
```
I wonder if it's worth reusing the string?
```suggestion
          R"(A Redpanda Enterprise Edition license is required to use "
          "enterprise feature "{}". This property is being ignored.)",
          config::shard_local_cfg().default_leaders_preference.name());

BenPope commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -848,6 +848,18 @@ leader_balancer::build_preference_index() {
         if (preference.has_value()) {
+            if (
+              should_sanction
+              && preference.value() != config::leaders_preference{}) {
```
I wonder if we should use the default value?
```suggestion
              && preference.value()
                   != config::shard_local_cfg()
                        .default_leaders_preference.default_value()) {
```

IoannisRP replied:
Once we are done with this work, this logic will be duplicated in here, in `feature_manager::report_enterprise_features` and in `get_enterprise_features` (both overloads).

We currently have the logic of what is a restricted value in the `enterprise` property. Given that this is a topic property that refers to the same thing, it feels like we may need to re-evaluate our design down the line, when we have seen more use cases. (Note that we have enterprise features that reflect node config, as well + some enterprise features don't match the reporting we want to be doing)

My gut feeling is that we should use the same notation in all of them, so it is less probable to forget about them if we get to refactor this. If we want to use the property instead of the pre-existing notation, i am happy to change all of the above to
```
if (config::shard_local_cfg().default_leaders_preference.check_restricted(value) {...}
```

BenPope replied:
> If we want to use the property instead of the pre-existing notation, i am happy to change all of the above to

Ahh, `check_restricted` feels right to me, too.

IoannisRP replied:
ok, i'll change it everywhere (and it will be easier to find all the instances if we come around to refactor) :+1:

BenPope commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -848,6 +848,18 @@ leader_balancer::build_preference_index() {
+              && preference.value() != config::leaders_preference{}) {
+                vlog(
+                  clusterlog.warn,
+                  "A Redpanda Enterprise Edition license is required to use "
+                  "topic property \"redpanda.leaders.preference\" in topic {}. "
+                  "This property is being ignored.",
```
I wonder if we should reuse the string?
```suggestion
                  R"(A Redpanda Enterprise Edition license is required to use "
                  "topic property "{}" in topic {}. This property is being ignored.)",
                  kafka::topic_property_leaders_preference,
```

BenPope replied:
....except that ends up with a cyclic dependency with bazel.

IoannisRP replied:
while i agree that using the variable name is better, are we good to include a kafka header? It feels off.

BenPope replied:
> are we good to include a kafka header? It feels off.

Well, not as it currently is, as bazel will complain.

In general, cluster module does too much stuff, but splitting the dependencies here isn't worth it.

IoannisRP commented:
- Replaced hardcoded property name with `config::property::name()` call
- Replaced ad-hoc checking for restricted `leadership_preference` value with call to `enterprise::check_restricted`

michael-redpanda commented:
@ztlpn would you mind taking a look at the sanctioning logic here and confirm that this won't cause any issues?  Goal of the sanction is that, when the license has expired, we ignore the cluster config and any topic specific properties related to leadership pinning.  Thanks!

michael-redpanda reviewed (commented):
Looks fine but would like @ztlpn to confirm the logic

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/57899-->
<!--5da0ba56-e549-4585-9e60-bce6cca37583-->
<!--title-->
## Retry command for [Build#57899](https://buildkite.com/redpanda/redpanda/builds/57899)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/follower_fetching_test.py::FollowerFetchingTest.test_basic_follower_fetching@{"read_from_object_store":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":null,"client_auth":true,"enable_authz":true,"use_sasl":true,"use_tls":true}
tests/rptest/tests/acls_test.py::AccessControlListTestUpgrade.test_describe_acls@{"authn_method":"none","client_auth":true,"enable_authz":true,"use_sasl":true,"use_tls":true}
tests/rptest/tests/archival_test.py::ArchivalTest.test_one_node_reconnect@{"cloud_storage_type":2}
tests/rptest/tests/rpk_registry_test.py::RpkRegistryTest.test_produce_consume_proto
tests/rptest/tests/shadow_indexing_admin_api_test.py::SIAdminApiTest.test_bucket_validation@{"cloud_storage_type":2}
```
<!--/body-->

ztlpn commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -821,8 +824,19 @@ leader_balancer_types::preference_index
 leader_balancer::build_preference_index() {
     leader_balancer_types::preference_index ret;
 
+    const bool should_sanction = _feature_table.should_sanction();
```
I wonder why we can't just add this condition to `leadership_pinning_enabled()` and call it a day? Is it only to get nice warning messages? But the user is already getting license nag messages from `feature_manager`...

IoannisRP replied:
> But the user is already getting license nag messages from feature_manager...

One of the design goals was to be as loud as possible with all the sanctionings, to minimise the possibility of someone trying to use an enterprise feature without a license, seeing that "it doesn't work" and thinking that it "just doesn't work" instead of understanding that this is an imposed restriction.

IoannisRP replied:
@ztlpn we have 2 functions called `leadership_pinning_enabled`. One in "leader_balancer.cc" and one in "feature_manager.cc". And they are checking for different things. I guess that not checking for the preferences in "leader_balancer.cc" might be an optimisation, but shouldn't "feature_manager.cc" also check the  `_enable_rack_awareness` value?

ztlpn replied:
Semantics are a bit different, in feature_manager it is "has the user configured the feature", in leader_balancer it is "should redpanda do anything about it". `_enable_rack_awareness` here acts as a kill switch for all rack-related features, and the configuration when leadership pinning is configured, but _enable_rack_awareness is false doesn't make sense outside of incidents. A nice usability improvement would be to print some warning in this case, but it is out of scope here.

I propose not to complicate logic here and sanction and nag the user regardless of the value of `_enable_rack_awareness`.

ztlpn commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -821,8 +824,19 @@ leader_balancer_types::preference_index
 leader_balancer::build_preference_index() {
     leader_balancer_types::preference_index ret;
 
+    const bool should_sanction = _feature_table.should_sanction();
```
We should also add a skipping condition in the `handle_topic_deltas()` function (line 197) and to `on_default_preference_changed()`, or the balancer will get unnecessary extra wakeups if the user changes those properties.

IoannisRP commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -442,9 +446,55 @@ bool leader_balancer::should_stop_balance() const {
     return _enable_rack_awareness();
 }
 
+std::optional<ss::sstring>
+leader_balancer::leadership_pinning_sanction_warning_msg() const {
+    const auto& [_, config_prop_sanctioned] = _default_preference(true);
+    chunked_vector<model::topic_namespace> sanctioned_topics;
+
+    for (const auto& topic : _topics.topics_map()) {
+        const auto& preference
+          = topic.second.get_configuration().properties.leaders_preference;
+        if (
+          preference.has_value()
+          && config::shard_local_cfg()
+               .default_leaders_preference.check_restricted(
+                 preference.value())) {
+            sanctioned_topics.push_back(topic.first);
+        }
+    }
```
There is a logic overlap between this function and the `feature_table::leadership_pinning_enabled`. It is worth addressing in a follow up refactoring, but i think this should be enough for this iteration.

ztlpn commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -442,9 +446,55 @@ bool leader_balancer::should_stop_balance() const {
+    if (!sanctioned_topics.empty()) {
+        msg += ssx::sformat(
+          " Value of topic property \"redpanda.leaders.preference\""
+          " in topics \"{}\" is being ignored.",
```
nit: probably better to limit the number of printed topics to ~10 to avoid absurdly long log messages.

michael-redpanda replied:
If this is a situation where we could cause an overallocation warning I'd be fine with limiting it.  If it's just because "it's annoying", then I'd say leave it.  The sanctions are there to be annoying :)

IoannisRP replied:
the way it is implemented, it could cause an overallocation, as `ss::sstring` seems to be contiguous. I'll have a look to see if this is easy to work around, otherwise i will limit it.

michael-redpanda commented on src/v/cluster/scheduling/leader_balancer.cc:
```diff
@@ -821,8 +871,13 @@ leader_balancer_types::preference_index
 leader_balancer::build_preference_index() {
     leader_balancer_types::preference_index ret;
 
+    vassert(
+      !_feature_table.should_sanction(),
+      "this path should be unreachable on sanctioned execution");
```
👍

IoannisRP commented:
- removed separate logging for the different settings that comprise the leadership pinning. Instead there is a single message if any of them is enabled.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/57961-->
<!--d7cff0f0-696b-431a-b6ad-9023bbeb3bbe-->
<!--title-->
## Retry command for [Build#57961](https://buildkite.com/redpanda/redpanda/builds/57961)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndHydrationTimeoutTest.test_hydration_completes_on_timeout
```
<!--/body-->

IoannisRP commented:
- Changed `should_print_warning` to use `ranges::any_of`
